File: spring-web-java/src/main/java/spring/web/java/domain/order/OrderItem.java
package spring.web.java.domain.order;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import spring.web.java.domain.Base;
import spring.web.java.domain.item.Item;

@Entity
@Getter
@Table(name = "order_item")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class OrderItem extends Base {

    @Id
    @GeneratedValue
    @Column(name = "order_item_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    private Item item;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    private Order order;

    private int orderPrice;
    private int count;

    public static OrderItem create(Item item, int orderPrice, int count) {
        OrderItem orderItem = new OrderItem();

        orderItem.item = item;
        orderItem.orderPrice = orderPrice;
        orderItem.count = count;

        item.removeQuantity(count);

        return orderItem;
    }

    public void setOrder(Order order) {
        this.order = order;
    }

    public void cancel() {
        item.addQuantity(count);
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/order/OrderItem.kt
package spring.web.kotlin.domain.order

import jakarta.persistence.*
import spring.web.kotlin.domain.Base
import spring.web.kotlin.domain.item.Item

@Entity
@Table(name = "order_item")
class OrderItem protected constructor(
    @Id
    @GeneratedValue
    @Column(name = "order_item_id")
    val id: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    val item: Item,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    var order: Order? = null,

    val orderPrice: Int,
    val count: Int
) : Base() {
    companion object {
        fun create(item: Item, orderPrice: Int, count: Int) = OrderItem(
            item = item,
            orderPrice = orderPrice,
            count = count
        )
            .also { it.item.removeQuantity(count) }
    }
}